### PR TITLE
JDK-8319307: DCmds should not assert on truncation and should report truncation

### PR DESCRIPTION
**Pull request by tstuefe**
`bufferedStream` was originally intended to provide intermediate output buffering; the buffer was supposed to drain upon flushing. However, in many cases, `flush()` is a no-op, so the buffer never gets drained.

To prevent infinitely raising buffer sizes for non-flushing `bufferedStream`, [JDK-8220394](https://bugs.openjdk.org/browse/JDK-8220394) introduced the notion of a "maximum reasonable cap". Upon reaching this threshold, we assert in debug VMs since we assumed this to be a condition worth analyzing. In release VMs, we silently truncate.

But DCmds - one of the primary users of `bufferedStream` - can reach the maximum cap under normal conditions; one example would be printing the list of dynamic libraries on Linux (just prints the process memory map) - this can get very large. Similarly, NMT detail reports and VM.info output can get just as large.

Therefore, neither asserting nor silent truncation is optimal. Instead, we should truncate the output, print a visible truncation marker, and - if possible - interrupt the printing.

---

The patch is minimally invasive to simplify review. Like most stream classes, `bufferedStream` would benefit from an overhaul, but I'd like to leave that to a future RFE.

Testing: Tested manually with a number of commands with artificially increased output size. GHAs (Windows test errors unrelated).

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blockers
&nbsp;⚠️ Too few reviewers with at least role reviewer found (have 0, need at least 1) (failed with the updated jcheck configuration)
&nbsp;⚠️ Whitespace errors (failed with the updated jcheck configuration)

### Issue
 * [JDK-8319307](https://bugs.openjdk.org/browse/JDK-8319307): DCmds should not assert on truncation and should report truncation (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16474/head:pull/16474` \
`$ git checkout pull/16474`

Update a local copy of the PR: \
`$ git checkout pull/16474` \
`$ git pull https://git.openjdk.org/jdk.git pull/16474/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16474`

View PR using the GUI difftool: \
`$ git pr show -t 16474`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16474.diff">https://git.openjdk.org/jdk/pull/16474.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16474#issuecomment-1792344567)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
I don't think I agree with this.

If DCmds are regularly asserting, or being truncated, then we should be using a larger buffer shouldn't we? That is what the assert was intended to be used for.

Having clients of the stream actively check for truncation also seems wrong to me. The basic model of our `bufferedStream` is that it is presumed large enough for all usecases (hence the assert when not the case) and users don't have to care. Otherwise the normal API approach here would be for `write()` to return a value indicating truncation so that the caller could consider resizing the buffer.

**Comment by tstuefe** [comment]
> If DCmds are regularly asserting, or being truncated, then we should be using a larger buffer shouldn't we? That is what the assert was intended to be used for.

The original bufferedStream did not have a limit. We added a limit with JDK-8220394 because we had cases where VMs were OOM-killed due to an error in printing code. But even beyond errors in printing code, a limit seems reasonable. E.g. when running with ZGC, I had VM.info reports that were larger than 2 GB due to the fragmented memory mapping. Do we want to cache that all?

If you have a limit, all the rest follows:
- it makes sense to signal the truncation to the caller
- it makes sense (though it is optional) to stop printing once the bufferedStream cannot receive the output anymore.

If we don't want a limit, we can either restore bufferedStream to its original state pre-JDK8220394 and live with the danger. Or maybe we can think about caching large output differently. Possibly writing temporary files to /tmp. Ideally, the jcmd communication protocol would allow streaming the output back to jcmd, but that is harder to do, especially not if one wants to remain compatible across JDK versions.

> Having clients of the stream actively check for truncation also seems wrong to me. The basic model of our bufferedStream is that it is presumed large enough for all usecases (hence the assert when not the case) and users don't have to care. Otherwise the normal API approach here would be for write() to return a value indicating truncation so that the caller could consider resizing the buffer.

The truncation check is optional, but reasonable IMHO, see ratio above.

> I don't think clients should be aware of, or worrying about truncation, this way. It seems very intrusive.

Hmm, I'd have thought you agree with this change since when discussing JDK-8220394 you proposed a truncation message:

https://mail.openjdk.org/pipermail/hotspot-runtime-dev/2019-May/034431.html

**Comment by dholmes-ora** [comment]
@tstuefe  thanks for the history lesson. So it was assumed writes can't fail and messages can't be truncated, which was true until we hit actual system limits. So we imposed a cap that we didn't expect to reach in practice (hence assert) and an indicator that the message had been truncated (which is a good thing). So we were still pretending writes can't fail.

Now you are saying, lets drop the pretense - the buffer is limited in size and callers have to (or can if they want to) deal with that. The question then is how? The usual way would be for writes to fail, report an error and caller then does something to resize the buffer or abort the writing. That would be very disruptive to the existing code. Instead you've added a way for a failing write to record that it failed in a side-channel: is_truncated(). The callers can then query this side-channel and decide what to do. Now that I write that all out this isn't as bad as I initially thought, but I still find it somewhat grating from an API perspective.

An alternative, for these problem cases, would be to go back to a buffer that attempts to grow as needed until it hits system limits. Are the actual buffers involved isolated so that we can choose which one(s) to change to be unlimited? If so are these commands ones that would likely lead to the OOM issues that caused the limit to be imposed in the first place?

**Comment by tstuefe** [comment]
> @tstuefe thanks for the history lesson. So it was assumed writes can't fail and messages can't be truncated, which was true until we hit actual system limits. So we imposed a cap that we didn't expect to reach in practice (hence assert) and an indicator that the message had been truncated (which is a good thing). So we were still pretending writes can't fail.

Yes

> 
> Now you are saying, lets drop the pretense - the buffer is limited in size and callers have to (or can if they want to) deal with that. The question then is how? The usual way would be for writes to fail, report an error and caller then does something to resize the buffer or abort the writing. That would be very disruptive to the existing code. Instead you've added a way for a failing write to record that it failed in a side-channel: is_truncated(). The callers can then query this side-channel and decide what to do. Now that I write that all out this isn't as bad as I initially thought, but I still find it somewhat grating from an API perspective.

Yes. Minus the grating, I think having an observable "buffer full" signal would make sense at least for stringStream, which can be run over an existing static buffer. I agree that bufferedStream feels more like it should not limit output but flush occasionally. IMHO the real error is not honoring the flush() request.

> 
> An alternative, for these problem cases, would be to go back to a buffer that attempts to grow as needed until it hits system limits. Are the actual buffers involved isolated so that we can choose which one(s) to change to be unlimited? If so are these commands ones that would likely lead to the OOM issues that caused the limit to be imposed in the first place?

Well, yes. Its the one buffer in the attach listener. We could just crank up the buffer maximum; we could also do that on a per-command base, for commands that we know may print a lot of output.

**Comment by tstuefe** [comment]
> > @tstuefe thanks for the history lesson. So it was assumed writes can't fail and messages can't be truncated, which was true until we hit actual system limits. So we imposed a cap that we didn't expect to reach in practice (hence assert) and an indicator that the message had been truncated (which is a good thing). So we were still pretending writes can't fail.
> 
> Yes
> 
> > Now you are saying, lets drop the pretense - the buffer is limited in size and callers have to (or can if they want to) deal with that. The question then is how? The usual way would be for writes to fail, report an error and caller then does something to resize the buffer or abort the writing. That would be very disruptive to the existing code. Instead you've added a way for a failing write to record that it failed in a side-channel: is_truncated(). The callers can then query this side-channel and decide what to do. Now that I write that all out this isn't as bad as I initially thought, but I still find it somewhat grating from an API perspective.
> 
> Yes. Minus the grating, I think having an observable "buffer full" signal would make sense at least for stringStream, which can be run over an existing static buffer. I agree that bufferedStream feels more like it should not limit output but flush occasionally. IMHO the real error is not honoring the flush() request.
> 
> > An alternative, for these problem cases, would be to go back to a buffer that attempts to grow as needed until it hits system limits. Are the actual buffers involved isolated so that we can choose which one(s) to change to be unlimited? If so are these commands ones that would likely lead to the OOM issues that caused the limit to be imposed in the first place?
> 
> Well, yes. Its the one buffer in the attach listener. We could just crank up the buffer maximum; we could also do that on a per-command base, for commands that we know may print a lot of output.

BTW, one possible improvement for bufferedStream could be to switch to mmap if backing buffer grows beyond a certain threshold. The benefit would be that upon release, this memory is really gone and does not lurk around increasing libc memory retention.

Another improvement could be to then reallocate by attempting to map adjacent pages. That removes the need for copying. One could even just use a VirtualSpace from the start, with a larger reserved size and commit on growth.

**Comment by dholmes-ora** [comment]
> IMHO the real error is not honoring the flush() request.

Did we look into fixing that?

**Comment by tstuefe** [comment]
> > IMHO the real error is not honoring the flush() request.
> 
> Did we look into fixing that?

No easy way to do that. The underlying jcmd protocol works on a request-answer base, so it is not prepared to stream output as it comes. Even if we change that, and it would be good to change, this would possibly add incompatibilities with older jcmd versions (being able to use any jcmd on any VM is really handy).

**Comment by tstuefe** [comment]
Hi @dholmes-ora, how should we proceed here?

- revert back to the earliest form (allocate as much as you need without limit)
- keep the new version (allocate until limit, then truncate with a warning)

The only thing I would like to get rid of ís the current behavior (allocate till limit, then either assert or truncate silently). I think that is actually the worst option.

**Comment by dholmes-ora** [comment]
> Well, yes. Its the one buffer in the attach listener. We could just crank up the buffer maximum; we could also do that on a per-command base, for commands that we know may print a lot of output.

Isn't this a third option?

**Comment by kevinjwalls** [comment]
Hi,
What jcmds are hitting the limit?  Dynamic library list on Linux is mentioned in the bug, so is that a "jcmd PID VM.dynlibs" command which is printing 100MB of output? (/proc/pid/maps is >100MB)

Is the problem that the buffer has nowhere to drain to at the point the diagnostic command is running?   AttachListener executes the whole diagnostic command into a buffer, then calls op->complete(...); which prints the result using e.g. LinuxAttachListener::write_fully().

If this no longer works, maybe it's time to use a version of bufferedStream which has knowledge of where output is going.  The attachlistener creates the buffer and tells it the destination (the attach listener's socket) so it can actually write it out and clear the buffer when necessary.  The AttachListener's call to op->complete would just flush() the buffer.

If it's only os::print_dll_info(outputStream *st) which is known to fill the buffer, then that one function could limit its own output until there is such a better solution.

**Comment by tstuefe** [comment]
> Hi, What jcmds are hitting the limit? Dynamic library list on Linux is mentioned in the bug, so is that a "jcmd PID VM.dynlibs" command which is printing 100MB of output? (/proc/pid/maps is >100MB)

at least the following commands are affected:

VM.dynlibs
VM.info, which prints the process map
System.map, which prints an annotated memory map
VM.native_memory, which in detail mode can get really large

the following command may be affected with high output details:
VM.metaspace

> 
> Is the problem that the buffer has nowhere to drain to at the point the diagnostic command is running? AttachListener executes the whole diagnostic command into a buffer, then calls op->complete(...); which prints the result using e.g. LinuxAttachListener::write_fully().

Yes and yes

> 
> If this no longer works, maybe it's time to use a version of bufferedStream which has knowledge of where output is going. The attachlistener creates the buffer and tells it the destination (the attach listener's socket) so it can actually write it out and clear the buffer when necessary. The AttachListener's call to op->complete would just flush() the buffer.

Yes, already tracked by https://bugs.openjdk.org/browse/JDK-8319055 . It is out of scope for this issue though, and I have no time for it; also, there are backward compatibility issues to be considered. See JBS issue for details.

> 
> If it's only os::print_dll_info(outputStream *st) which is known to fill the buffer, then that one function could limit its own output until there is such a better solution.

`os::print_dll_info` should not have any knowledge of the context it is used in and of the limitations of the stream handed in. That's its caller's business. Separation of concerns.

The solution would be to go forward with this PR. It is a trivial change, and the current state is unacceptable. Unfortunately, this PR seems to be blocked dead.

**Comment by tstuefe** [comment]
> > Well, yes. Its the one buffer in the attach listener. We could just crank up the buffer maximum; we could also do that on a per-command base, for commands that we know may print a lot of output.
> 
> Isn't this a third option?

No, because it does not change the fact that if a command reaches the limit - be it global or per-command - we still have the same behavior (assert or silent truncation)

**Comment by kevinjwalls** [comment]
Thanks, JDK-8319055 does indeed need more investigation.

While nobody would want to _continually_ raise the buffer size, raising it is the easy one-liner with no side-effects on multiple diagnostic commands...  Would it be 200MB?  500MB? 8-)

**Comment by tstuefe** [comment]
> Thanks, JDK-8319055 does indeed need more investigation.
> 
> While nobody would want to _continually_ raise the buffer size, raising it is the easy one-liner with no side-effects on multiple diagnostic commands... Would it be 200MB? 500MB? 8-)
> 
> (sorry, should have written something like "multiple other files", not diagnostic commands)

Raising the limit is unrelated to this PR. We can certainly raise the limit, but it is not possible to prevent the limit from being hit during normal operations. Or more precise, you can prevent it but you would have to set it so high that it's basically useless for preventing most OOM kills, especially in the cloud. Which was the original intent of this limit. As an example, we have seen process maps > 2GB for large processes with very fragmented address spaces.

What this PR tries to change is how the VM *reacts* to the limit, regardless of where the threshold is. IMHO the VM should indicate that it hit a limit. It should not silently truncate the output. Nor should it assert.

File: src/hotspot/os/aix/attachListener_aix.cpp [hunk 1]
```diff
@@ -451,6 +451,10 @@ void AixAttachOperation::complete(jint result, bufferedStream* st) {
     // Shutdown the socket in the cleanup function to enable more than
     // one agent attach in a sequence (see comments to listener_cleanup()).
     AixAttachListener::write_fully(this->socket(), (char*) st->base(), st->size());
+    if (st->was_truncated()) {
+      char msg [] = "\n(truncated)\n";
+      AixAttachListener::write_fully(this->socket(), msg, strlen(msg));
+    }
```
Comment (dholmes-ora) [inline src/hotspot/os/aix/attachListener_aix.cpp:457]: Isn't this happening too late? The st content has already been truncated and we have written the "final" message to the socket for the client to read, and then we try to send a follow-up message that the original was truncated. ???

Comment (dholmes-ora) [reply src/hotspot/os/aix/attachListener_aix.cpp:457]: Okay. Do we not then have to check the rc of the first `write_fully` and only do this follow up write if `rc == 0`?

File: src/hotspot/os/linux/os_linux.cpp [hunk 2]
```diff
@@ -1889,7 +1889,7 @@ static bool _print_ascii_file(const char* filename, outputStream* st, unsigned*
   ssize_t bytes;
   buf[32] = '\0';
   unsigned lines = 0;
-  while ((bytes = ::read(fd, buf, sizeof(buf)-1)) > 0) {
+  while (!st->was_truncated() && (bytes = ::read(fd, buf, sizeof(buf)-1)) > 0) {
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:1892]: Okay so we stop printing if `st` is full but where is the notification that this has happened?

File: src/hotspot/share/nmt/memReporter.cpp [hunk 3]
```diff
@@ -320,7 +320,7 @@ int MemDetailReporter::report_malloc_sites() {
 
   const MallocSite* malloc_site;
   int num_omitted = 0;
-  while ((malloc_site = malloc_itr.next()) != nullptr) {
+  while (!output()->was_truncated() && (malloc_site = malloc_itr.next()) != nullptr) {
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/nmt/memReporter.cpp:323]: Again where do we inform the consumer of the stream that the content has been truncated?

File: src/hotspot/share/utilities/ostream.hpp [hunk 4]
```diff
@@ -299,7 +302,9 @@ class bufferedStream : public outputStream {
   bufferedStream(size_t initial_bufsize = 256, size_t bufmax = 1024*1024*10);
   bufferedStream(char* fixed_buffer, size_t fixed_buffer_size, size_t bufmax = 1024*1024*10);
   ~bufferedStream();
-  virtual void write(const char* c, size_t len);
+  virtual void write(const char* c, size_t len) override;
+  bool was_truncated() const override { return truncated; }
+  // Returns number of characters written, exclusing terminating zero.
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/ostream.hpp:307]: typo: exclusing -> excluding the

File: src/hotspot/share/nmt/memReporter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/memReporter.cpp:323]: Delayed, at the caller. NMT report is called via dcmd, and attach listener queries the truncation flag. Here, we only query the truncation flag to save unnecessary work.

File: src/hotspot/os/linux/os_linux.cpp (hunk 2 above)
Comment (tstuefe) [reply src/hotspot/os/linux/os_linux.cpp:1892]: Caller needs to take care of it.

In my mind, the caller knows what stream class it uses, so it knows if the stream is one that can truncate. So its his responsibility to handle stream::is_truncated. In this case, the caller that uses a truncatable stream is the dcmd framework, and it handles this.

The printing routines themselves could be completely oblivious to it, I only query truncated in a handful of places to save work - if we know the output stream truncated, there is no reason to continue work printing (and potentially, holding a lock while doing it).

Comment (dholmes-ora) [reply src/hotspot/os/linux/os_linux.cpp:1892]: Okay, so it isn't clear to me that all of these situations are such that the caller is always a DCmd and that it always reports the truncation due to the attach code changes.

**Comment by kevinjwalls** [comment]
> Raising the limit is unrelated to this PR. 

But they are related.  Raising the limit is the alternative to this PR, for most users most of the time.
I think it was reasonable to ask what kind of size are you seeing, and with what kind of frequency.
That should help say whether this is a bump the limit situation, or a we-must-fix-this-now situation.

JDK-8319055 says we should rework this in a better way, which I expect we all agree on, and this current PR is a possible solution until that change comes along (which might be undone by JDK-8319055).

/proc/pid/maps was > 2GB of text?  At that point nobody can help you. 8-)
If that maps list ended before showing the high numbers, nobody would be surprised!

**Comment by tstuefe** [comment]
> > Raising the limit is unrelated to this PR.
> 
> But they are related. Raising the limit is the alternative to this PR, for most users most of the time. I think it was reasonable to ask what kind of size are you seeing, and with what kind of frequency.

Oh, certainly.

> That should help say whether this is a bump the limit situation, or a we-must-fix-this-now situation.

Sure. I am just surprised how contentious the printing of a simple "truncated" message turns out to be. Unfortunately, I am out of time for this PR. Maybe it is just better to leave this unresolved.

> 
> JDK-8319055 says we should rework this in a better way, which I expect we all agree on, and this current PR is a possible solution until that change comes along (which might be undone by JDK-8319055).
> 
> /proc/pid/maps was > 2GB of text? At that point nobody can help you. 8-) 

That was with ZGC, Oracle's own collector, and a 4-5TB heap if I recall correctly. That was a real customer case. The number of mappings relate to the ZUnmapper thread being raced by mutator threads. I think they work on a solution for this, but my point is that this could happen.

>If that maps list ended before showing the high numbers, nobody would be surprised!

I am concerned about truncation that is not immediately obvious; also, when doing support, truncation may happen at other levels (e.g. customer uploads a file to a support web site which truncates the output).

File: src/hotspot/os/linux/os_linux.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/os/linux/os_linux.cpp:1892]: Okay.

**Comment by dholmes-ora** [comment]
My main concern here was that if are going to stop pretending we have an unlimited buffer, then the usual API design here would be for `write` to report an error. But because we don't want to have to change all users of this to deal with such an API change, the `full` state is being communicated via a side-channel: `is_truncated()`. As I wrote above I don't really like this but I have come around and agree that this is better than the status quo for the problematic situation with the Dcmd. But other clients now only get silent truncation, without the benefit of the assertion failure to report an unexpected too-small buffer. This latter aspect seems a backward step to me as we won't know we have a problem somewhere. Even if we report truncation in all cases we still may not notice that we need a bigger buffer until we get bug reports from the field about the truncation.

I guess a key thing here is that we need to know when we are filling a buffer with a known finite-size information block, and when that information could have an arbitrary size (that would overflow the buffer). At the moment we simply hope one-max-size-fits-all and rely on the assert to tell us otherwise - not that that would catch the reported customer issue with the multi TB heap and the massive process map table.

**Pull request closed without merging**